File: 回溯/78.子集.java
class Solution {
    List<List<Integer>> res;
    List<Integer> path;
    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        path = new ArrayList<>();
        dfs(nums,0);
        return res;
    }

    public void dfs(int[] nums,int start){
        if(start == nums.length){
            res.add(new ArrayList<>(path));
            return ;
        }
        dfs(nums,start+1);
        path.add(nums[start]);
        dfs(nums,start+1);
        path.removeLast();
    }
}

File: 回溯/78.子集.go
func subsets(nums []int) [][]int {
	res := make([][]int, 0)
	path := make([]int, 0)

	var dfs func(int)
	dfs = func(start int) {
		if start == len(nums) {
			tmp := make([]int, len(path))
			for i, v := range path {
				tmp[i] = v
			}
			res = append(res, tmp)
			return
		}
		dfs(start + 1)
		path = append(path, nums[start])
		dfs(start + 1)
		path = path[:len(path)-1]
	}
	dfs(0)
	return res

}